test/shared/src/main/scala/zio/test/ZIOSpec.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2021 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import org.portablescala.reflect.annotation.EnableReflectiveInstantiation
+import zio._
+import zio.test.environment.TestEnvironment
+import zio.test.render._
+
+@EnableReflectiveInstantiation
+abstract class ZIOSpec[R <: Has[_]: Tag](val layer: ZLayer[TestEnvironment, Any, R]) extends ZIOApp { self =>
```

Comment:
Ideally would like to unify over the case where the environment is `Any`. May be able to use `Has.Union` to do this.

Reply:
May just be easier to have `DefaultZIOSpec` and `ZIOSpec` here.

Reply:
I think this `val layer` could as well be a `def`, since we are using `def` style for `spec` and it's odd to mix the two styles.

I agree on `ZIOSpec` and `ZIOSpecDefault`. I'd even suggest that the latter _extend_ the former, so everything can be treated as a `ZIOSpec`.

Reply:
Done.